src/librustc_typeck/check/method/mod.rs

```diff
@@ -290,7 +290,7 @@ impl<'a, 'gcx, 'tcx> FnCtxt<'a, 'gcx, 'tcx> {
         // type parameters or early-bound regions.
         let tcx = self.tcx;
         let method_item =
-            self.associated_item(trait_def_id, m_name, Namespace::Value).unwrap();
+            self.associated_item(trait_def_id, m_name, Namespace::Value)?;
```

Comment:
I think we should `tcx.sess.delay_span_bug` here in case of `None`. Returning `None` will then require some regular compilation error to happen later.

Reply:
Ok, this is my first rustc commit, so maybe you or someone can point me in the right direction, where something similar is implemented? and what exactly delay_span_bug does?
Currently the message in the test case presented in #31076 is "kind of right", i.e. user implements add trait incorrectly and message says that trait is not implemented instead of panicking (as it is now), but it cloud be better, how ever this here seems to be a regular compilation error to me - user gave incorrect source, rustc returned error message that is 90% relevant

Reply:
Sorry, you're right. I should have explained a bit.

`delay_span_bug` pushes an error to a an error list that is only uses in case no other errors have been issued.

It basically ensures that there will be an error, even if users manage to hit your `None` + `?` case without triggering the error seen in the issue. Think of it as a canary. We are emitting an error that is only shown if no better error shows up.

I think in this case you can do

```rust
let method_item = match self.associated_item(trait_def_id, m_name, Namespace::Value) {
    Some(method_item) => method_item,
    None => {
        tcx.sess.delay_span_bug(span, "operator trait does not have corresponding operator method);
        return None;
    }
};
```

Reply:
No reason to be sorry.
I will test this locally and update the pull request to do as advised.

Thank you!

Reply:
added the changes as advised by @oli-obk ,
nitpick: is formatting ok like this for the new commit? or should i store the message str and use that in the function call?

Reply:
Looks ok to me. We're not employing rustfmt on this repo yet, but there are plans for that.